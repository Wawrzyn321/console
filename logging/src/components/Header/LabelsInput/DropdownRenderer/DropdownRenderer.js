import React from 'react';
import PropTypes from 'prop-types';
import './DropdownRenderer.scss';

DropdownRenderer.propTypes = {
  recentLabels: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
  logLabels: PropTypes.arrayOf(PropTypes.object.isRequired).isRequired,
  chooseLabel: PropTypes.func.isRequired,
};

export default function DropdownRenderer({
  recentLabels,
  logLabels,
  chooseLabel,
}) {
  const [statefulLogLabels, setStatefulLogLabels] = React.useState(
    logLabels
      .filter(logLabel => logLabel.labels.length)
      .map(logLabel => {
        return {
          isHidden: true,
          logLabel: logLabel,
        };
      }),
  );

  function formatName(name) {
    return name.replace(' ', '_').toLowerCase();
  }

  function switchState(labelName) {
    setStatefulLogLabels(
      statefulLogLabels.map(entry => {
        if (entry.logLabel.name === labelName) {
          return {
            ...entry,
            isHidden: !entry.isHidden,
          };
        } else {
          return {
            ...entry,
            isHidden: true,
          };
        }
      }),
    );
  }

  const recentLabelsList = recentLabels.length ? (
    <ul className="fd-mega-menu__list">
      {recentLabels.map(entry => (
        <li key={entry} className="fd-mega-menu__item">
          <span
            className="cursor-pointer fd-mega-menu__link"
            onClick={() => chooseLabel(entry)}
          >
            {entry}
          </span>
        </li>
      ))}
    </ul>
  ) : (
    <span
      className="fd-mega-menu__item--disabled"
      data-test-id="no-recent-labels"
    >
      No recent labels
    </span>
  );

  const logLabelsSubList = ({ logLabel, isHidden }) => (
    <ul
      className="fd-mega-menu__sublist"
      id={logLabel.name}
      aria-hidden={isHidden}
    >
      {logLabel.labels.map(name => (
        <li className="fd-mega-menu__subitem" key={name}>
          <span
            className="fd-mega-menu__sublink cursor-pointer"
            onClick={() =>
              chooseLabel(`${formatName(logLabel.name)}="${name}"`)
            }
          >
            {name}
          </span>
        </li>
      ))}
    </ul>
  );

  const logLabelsList = logLabels.length ? (
    <ul className="fd-mega-menu__list">
      {statefulLogLabels.map(l => (
        <li className="fd-mega-menu__item" key={l.logLabel.name}>
          <span
            className="fd-mega-menu__link has-child cursor-pointer"
            aria-controls={l.logLabel.name}
            aria-haspopup="true"
            onClick={() => switchState(l.logLabel.name)}
          >
            {l.logLabel.name}
          </span>
          {logLabelsSubList(l)}
        </li>
      ))}
    </ul>
  ) : (
    <span className="fd-mega-menu__item--disabled" data-test-id="no-log-labels">
      No log labels
    </span>
  );

  return (
    <nav className="fd-mega-menu">
      <div className="fd-mega-menu__group">
        <h3 className="fd-mega-menu__title">Recently Selected Labels</h3>
        {recentLabelsList}
      </div>
      <div className="fd-mega-menu__group">
        <h3 className="fd-mega-menu__title">Log Labels</h3>
        {logLabelsList}
      </div>
    </nav>
  );
}
