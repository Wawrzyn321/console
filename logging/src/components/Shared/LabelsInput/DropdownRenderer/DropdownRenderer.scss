.fd-mega-menu__item--disabled {
    padding: 10px 20px;
    color: #51555a;
    display: block;
    font-size: 1rem;
    line-height: 1.42857;
    font-weight: 400;
    color: var(--fd-color-text-4);
}
