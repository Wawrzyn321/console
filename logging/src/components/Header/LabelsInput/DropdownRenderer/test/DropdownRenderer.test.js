import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import DropdownRenderer from './../DropdownRenderer';

const mockRecentLabels = ['a', 'b'];

const mockLogLabels = [
  {
    name: 'name',
    labels: [['c', 'd']],
  },
];

describe('DropdownRenderer', () => {
  it('Renders with minimal props', () => {
    const component = renderer.create(
      <DropdownRenderer
        recentLabels={mockRecentLabels}
        logLabels={mockLogLabels}
        chooseLabel={() => {}}
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Shows "No log labels" text when there is no log labels', () => {
    const component = shallow(
      <DropdownRenderer
        recentLabels={mockRecentLabels}
        logLabels={[]}
        chooseLabel={() => {}}
      />,
    );
    expect(component.find('button[data-test-id="no-log-labels"]')).toBeTruthy();
  });

  it('Shows "No recent labels" text when there is no recent labels', () => {
    const component = shallow(
      <DropdownRenderer
        recentLabels={[]}
        logLabels={mockLogLabels}
        chooseLabel={() => {}}
      />,
    );
    expect(
      component.find('button[data-test-id="no-recent-labels"]'),
    ).toBeTruthy();
  });
});
