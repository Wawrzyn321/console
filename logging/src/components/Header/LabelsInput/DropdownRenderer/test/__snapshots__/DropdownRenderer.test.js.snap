// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownRenderer Renders with minimal props 1`] = `
<nav
  className="fd-mega-menu"
>
  <div
    className="fd-mega-menu__group"
  >
    <h3
      className="fd-mega-menu__title"
    >
      Recently Selected Labels
    </h3>
    <ul
      className="fd-mega-menu__list"
    >
      <li
        className="fd-mega-menu__item"
      >
        <span
          className="cursor-pointer fd-mega-menu__link"
          onClick={[Function]}
        >
          a
        </span>
      </li>
      <li
        className="fd-mega-menu__item"
      >
        <span
          className="cursor-pointer fd-mega-menu__link"
          onClick={[Function]}
        >
          b
        </span>
      </li>
    </ul>
  </div>
  <div
    className="fd-mega-menu__group"
  >
    <h3
      className="fd-mega-menu__title"
    >
      Log Labels
    </h3>
    <ul
      className="fd-mega-menu__list"
    >
      <li
        className="fd-mega-menu__item"
      >
        <span
          aria-controls="name"
          aria-haspopup="true"
          className="fd-mega-menu__link has-child cursor-pointer"
          onClick={[Function]}
        >
          name
        </span>
        <ul
          aria-hidden={true}
          className="fd-mega-menu__sublist"
          id="name"
        >
          <li
            className="fd-mega-menu__subitem"
          >
            <span
              className="fd-mega-menu__sublink cursor-pointer"
              onClick={[Function]}
            >
              c
              d
            </span>
          </li>
        </ul>
      </li>
    </ul>
  </div>
</nav>
`;
